test/mitmproxy/addons/test_savehar.py

```diff
@@ -0,0 +1,310 @@
+                    "path": "/",
+                    "domain": ".googls.com",
+                    "httpOnly": False,
+                    "secure": False,
+                    "sameSite": "none",
+                },
+            ],
+        ),
+    ],
+)
+def test_response_cookies(header: Headers, expected: list[dict]):
+    s = SaveHar()
+    resp = Response.make(200, "", header)
+    assert s.format_response_cookies(resp) == expected
+
+
+def test_seen_server_conn():
+    s = SaveHar()
+
+    flow = tflow.twebsocketflow()
+
+    servers_seen: set[Server] = set()
+    servers_seen.add(flow.server_conn)
+
+    calculated_timings = s.flow_entry(flow, servers_seen)["timings"]
+
+    assert calculated_timings["connect"] == -1.0
+    assert calculated_timings["ssl"] == -1.0
+
+
+def test_timestamp_end():
+    s = SaveHar()
+    servers_seen: set[Server] = set()
+    flow = tflow.twebsocketflow()
+
+    assert s.flow_entry(flow, set())["timings"]["send"] == 1
+
+    flow.request.timestamp_end = None
+    calculated_timings = s.flow_entry(flow, servers_seen)["timings"]
+
+    assert calculated_timings["send"] == 0
+
+
+def test_tls_setup():
+    s = SaveHar()
+    servers_seen: set[Server] = set()
+    flow = tflow.twebsocketflow()
+    flow.server_conn.timestamp_tls_setup = None
+
+    assert s.flow_entry(flow, servers_seen)["timings"]["ssl"] is None
+
+
+def test_binary_content():
+    s = SaveHar()
+
+    flow = tflow.twebsocketflow()
+    assert flow.response
+
+    flow.response.set_content(b"\xFF")
+    servers_seen: set[Server] = set()
+
+    assert flow.response.content
+
+    assert {
+        "size": 1,
+        "compression": 0,
+        "mimeType": "",
+        "text": "/w==",
+        "encoding": "base64",
+    } == s.flow_entry(flow, servers_seen)["response"]["content"]
+
+
+@pytest.mark.parametrize(
+    "log_file", [pytest.param(x, id=x.stem) for x in test_dir.glob("data/flows/*.mitm")]
+)
+def test_savehar(log_file: Path, tmp_path: Path, monkeypatch):
+    monkeypatch.setattr(version, "VERSION", "1.2.3")
+    s = SaveHar()
+
+    flows = io.read_flows_from_paths([log_file])
+
+    s.export_har(flows, types.Path(tmp_path / "testing_flow.har"))
+    expected_har = json.loads(
+        Path(test_dir / f"data/flows/{log_file.stem}.har").read_bytes()
+    )
+    actual_har = json.loads(Path(tmp_path / "testing_flow.har").read_bytes())
+
+    assert actual_har == expected_har
+
+
+def test_savehar_dump(tmpdir, tdata):
+    with taddons.context() as tctx:
+        a = tctx.script(tdata.path("../mitmproxy/addons/savehar.py"))
+        assert a
+        path = str(tmpdir.join("somefile"))
+        tctx.configure(a, hardump=path)
+
+        for x in io.read_flows_from_paths(
+            [Path(test_dir / "data/flows/websocket.mitm")]
+        ):
+            a.websocket_end(x)
+        a.done()
+
+        with open(path) as inp:
+            har = json.load(inp)
+            har["log"]["creator"]["comment"] = "mitmproxy version 1.2.3"
```

Comment:
Monkeypatch didnt work with the savehar dump approach, so I changed it manually for each case. Is there a better way to do this?

Reply:
You mean monkepatch did not work because the top-level `HAR` object was already created? An easy workaround here could be to only store the pages at runtime, and create the "outer frame" within `done`? :)